File: src/Runtime/BuiltInFunction.cs
/*
	PeakScriptin httpsin//github.com/Jenocn/PeakScript
	By Jenocnin httpsin//jenocn.github.io
*/

using System;
using System.Collections.Generic;

namespace peak.interpreter {

	public class BuiltInFunction {
		public static BuiltInFunction instance = new BuiltInFunction();
		public Dictionary<string, Variable> variables = new Dictionary<string, Variable>();

		public Variable FindVariable(string name) {
			Variable ret = null;
			if (variables.TryGetValue(name, out ret)) {
				return ret;
			}
			return null;
		}

		private BuiltInFunction() {
			Action<string, int, Func<List<Value>, Space, Value>> _Emplace = (string name, int paramSize, Func<List<Value>, Space, Value> func) => {
				var value = new ValueFunction(paramSize, func);
				var variable = new Variable(name, VariableAttribute.Const, value);
				variables.Add(name, variable);
			};

			// print
			_Emplace.Invoke("print", 1, (List<Value> args, Space s) => {
				string ret = "";
				foreach (var arg in args) {
					ret += ValueTool.ToString(arg);
				}
				System0.Echo(ret);
				return ValueNull.DEFAULT_VALUE;
			});
			// type
			_Emplace("type", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				return new ValueString(ValueTool.ToTypeString(args[0]));
			});
			// is_null
			_Emplace("is_null", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				bool ret = true;
				foreach (var arg in args) {
					ret &= ValueTool.IsNull(arg);
				}
				return new ValueBool(ret);
			});
			// is_number
			_Emplace("is_number", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				bool ret = true;
				foreach (var arg in args) {
					ret &= ValueTool.IsNumber(arg);
				}
				return new ValueBool(ret);
			});
			// is_bool
			_Emplace("is_bool", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				bool ret = true;
				foreach (var arg in args) {
					ret &= ValueTool.IsBool(arg);
				}
				return new ValueBool(ret);
			});
			// is_string
			_Emplace("is_string", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				bool ret = true;
				foreach (var arg in args) {
					ret &= ValueTool.IsString(arg);
				}
				return new ValueBool(ret);
			});
			// is_array
			_Emplace("is_array", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				bool ret = true;
				foreach (var arg in args) {
					ret &= ValueTool.IsArray(arg);
				}
				return new ValueBool(ret);
			});
			// is_function
			_Emplace("is_function", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				bool ret = true;
				foreach (var arg in args) {
					ret &= ValueTool.IsFunction(arg);
				}
				return new ValueBool(ret);
			});
			// is_object
			_Emplace("is_object", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				bool ret = true;
				foreach (var arg in args) {
					ret &= ValueTool.IsObject(arg);
				}
				return new ValueBool(ret);
			});
			// len
			_Emplace("len", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				var value = args[0];
				if (ValueTool.IsArray(value)) {
					return new ValueNumber((double) (value as ValueArray).value.Count);
				} else if (ValueTool.IsString(value)) {
					return new ValueNumber((double) (value as ValueString).value.Length);
				}
				return ValueNull.DEFAULT_VALUE;
			});
			// to_string
			_Emplace("to_string", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				var value = args[0];
				return new ValueString(ValueTool.ToString(value));
			});
			// to_number
			_Emplace("to_number", 1, (List<Value> args, Space s) => {
				if (args.Count == 0) {
					return null;
				}
				var value = args[0];
				if (!ValueTool.IsString(value)) {
					return null;
				}
				double ret = 0;
				if (!double.TryParse((value as ValueString).value, out ret)) {
					ret = 0;
				}
				return new ValueNumber(ret);
			});
		}
	}
}

File: src/runtime/builtin_function.cpp
#include "builtin_function.h"
#include "space.h"
#include "system.h"
#include "value/value_tool.h"
#include "variable.h"

using namespace peak;

BuiltInFunction* BuiltInFunction::GetInstance() {
	static BuiltInFunction _ins;
	return &_ins;
}

std::shared_ptr<Variable> BuiltInFunction::FindVariable(const std::string& name) const {
	auto ite = _variables.find(name);
	if (ite != _variables.end()) {
		return ite->second;
	}
	return nullptr;
}

BuiltInFunction::BuiltInFunction() {
	auto _Emplace = [this](const std::string& name, std::size_t paramSize, ValueFunction::FunctionType func) {
		auto value = std::make_shared<ValueFunction>(paramSize, func);
		auto variable = std::make_shared<Variable>(name, VariableAttribute::Const, value);
		_variables.emplace(name, variable);
	};

	// print
	_Emplace("print", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		std::string ret = "";
		for (auto arg : args) {
			ret += ValueTool::ToString(arg.get());
		}
		System::Echo(ret);
		return ValueNull::DEFAULT_VALUE;
	});
	// type
	_Emplace("type", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		return std::make_shared<ValueString>(ValueTool::ToTypeString(args[0].get()));
	});
	// is_null
	_Emplace("is_null", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		bool ret = true;
		for (auto arg : args) {
			ret &= ValueTool::IsNull(arg.get());
		}
		return std::make_shared<ValueBool>(ret);
	});
	// is_number
	_Emplace("is_number", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		bool ret = true;
		for (auto arg : args) {
			ret &= ValueTool::IsNumber(arg.get());
		}
		return std::make_shared<ValueBool>(ret);
	});
	// is_bool
	_Emplace("is_bool", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		bool ret = true;
		for (auto arg : args) {
			ret &= ValueTool::IsBool(arg.get());
		}
		return std::make_shared<ValueBool>(ret);
	});
	// is_string
	_Emplace("is_string", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		bool ret = true;
		for (auto arg : args) {
			ret &= ValueTool::IsString(arg.get());
		}
		return std::make_shared<ValueBool>(ret);
	});
	// is_array
	_Emplace("is_array", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		bool ret = true;
		for (auto arg : args) {
			ret &= ValueTool::IsArray(arg.get());
		}
		return std::make_shared<ValueBool>(ret);
	});
	// is_function
	_Emplace("is_function", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		bool ret = true;
		for (auto arg : args) {
			ret &= ValueTool::IsFunction(arg.get());
		}
		return std::make_shared<ValueBool>(ret);
	});
	// is_object
	_Emplace("is_object", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		bool ret = true;
		for (auto arg : args) {
			ret &= ValueTool::IsObject(arg.get());
		}
		return std::make_shared<ValueBool>(ret);
	});
	// len
	_Emplace("len", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		auto value = args[0];
		if (ValueTool::IsArray(value.get())) {
			return std::make_shared<ValueNumber>(static_cast<double>(std::static_pointer_cast<ValueArray>(value)->GetArray().size()));
		} else if (ValueTool::IsString(value.get())) {
			return std::make_shared<ValueNumber>(static_cast<double>(std::static_pointer_cast<ValueString>(value)->GetValue().length()));
		}
		return ValueNull::DEFAULT_VALUE;
	});
	// to_string
	_Emplace("to_string", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		auto value = args[0];
		return std::make_shared<ValueString>(ValueTool::ToString(value.get()));
	});
	// to_number
	_Emplace("to_number", 1, [](const std::vector<std::shared_ptr<Value>>& args, std::shared_ptr<Space>) -> std::shared_ptr<Value> {
		if (args.empty()) {
			return nullptr;
		}
		auto value = args[0];
		if (!ValueTool::IsString(value.get())) {
			return nullptr;
		}
		return std::make_shared<ValueNumber>(std::atof(std::static_pointer_cast<ValueString>(value)->GetValue().c_str()));
	});
}

